client.js

```diff
@@ -100,75 +112,94 @@ RingpopClient.prototype.destroy = function destroy(callback) {
-                channelType: 'topChannel'
+            callback(ClientRequestLimitError({
+                inflightCurrent: inflightCurrent,
+                inflightMax: inflightMax,
+                endpoint: endpoint
             }));
         });
         return;
     }
 
-    if (!opts || !validateHostPort(opts.host)) {
-        callback(InvalidHostPortError({
-            hostPort: String(opts && opts.host)
-        }));
-        return;
+    // Registering the request by its ID must happen at the precise
+    // moment before we know it will be "inflight."
+    this.requestsById[request.id] = request;
+    request.send();
+
+    function onSend() {
+        delete self.requestsById[request.id];
+        var args = Array.prototype.splice.call(arguments, 0);
+        callback.apply(null, args);
     }
+};
 
-    self.subChannel.waitForIdentified({
-        host: opts.host
-    }, function onIdentified(err) {
-        if (err) {
-            callback(err);
-            return;
+RingpopClient.prototype._scanForWedgedRequests = function _scanForWedgedRequests() {
+    var wedgedRequests = [];
+
+    // Find all requests that have not been timed-out by TChannel.
+    // Induce Ringpop timeout for wedged requests and cancel them.
+    var requestIds = Object.keys(this.requestsById);
+    for (var i = 0; i < requestIds.length; i++) {
+        var request = this.requestsById[requestIds[i]];
+        var timeInflight = Date.now() - request.timestamp;
+        var isWedged = timeInflight >= (this.config.get('wedgedRequestTimeout') ||
+            DEFAULT_WEDGED_REQUEST_TIMEOUT);
+        if (isWedged) {
+            delete this.requestsById[request.id];
+            request.cancel();
+            wedgedRequests.push(request);
         }
+    }
 
-        self.subChannel.request({
-            host: opts.host,
-            serviceName: 'ringpop',
-            hasNoParent: true,
-            retryLimit: opts.retryLimit || 0,
-            trace: false,
-            headers: {
-                as: 'raw',
-                cn: 'ringpop'
-            },
-            timeout: opts.timeout || 30000
-        }).send(endpoint, JSON.stringify(head), JSON.stringify(body), onSend);
-    });
-
-    function onSend(err, res, arg2, arg3) {
-        if (!err && !res.ok) {
-            err = safeParse(arg3) || new Error('Server Error');
-        }
+    if (wedgedRequests.length > 0) {
+        this.logger.error('ringpop canceled wedged requests', {
+            local: this.ringpop.whoami(),
+            numWedgedRequests: wedgedRequests.length,
+            wedgedRequests: wedgedRequestsToLog()
+        });
+    }
 
-        if (err) {
-            callback(err);
-            return;
-        }
+    return wedgedRequests;
 
-        callback(null, safeParse(arg3));
+    // Log the first N requests only. We don't want to flood
+    // our pipes.
+    function wedgedRequestsToLog() {
+        return _.take(wedgedRequests, 10).map(function toLog(request) {
```

Comment:
If we log less than what we actually do, can we log the total number, so that we have some record of actual execution?